File: ios/Classes/EcmpPlugin.swift
import Flutter
import UIKit
import ecommpaySDK

public class EcmpPlugin: NSObject, FlutterPlugin {
    
    private let ecompaySDK = EcommpaySDK()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "ecmpplugin", binaryMessenger: registrar.messenger())
        let instance = EcmpPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let json = call.arguments as? String else{
            result(FlutterMethodNotImplemented)
            return
        }
        switch call.method {
            case "sdkRun":
                sdkRun(json: json, result: result)
            case "getParamsForSignature":
                getParamsForSignature(json: json, result: result)
            default:
                result(FlutterMethodNotImplemented)
        }
    }
    
    private func getParamsForSignature(json: String, result: @escaping FlutterResult) {
        let jsonData = json.data(using: .utf8)!
        let paymentInfo: PluginPaymentInfo = try! JSONDecoder().decode(PluginPaymentInfo.self, from: jsonData)
        let paramsForSignature = paymentInfo.map().paramsForSignature
        result(paramsForSignature)
    }
    
    private func sdkRun(json: String, result: @escaping FlutterResult) {
        guard let controller = UIApplication.shared.delegate?.window??.rootViewController  as? FlutterViewController else {
            return
        }
        
        let jsonData = json.data(using: .utf8)!
        let pluginPaymentOptions: PluginPaymentOptions = try! JSONDecoder().decode(PluginPaymentOptions.self, from: jsonData)
        
        
        let paymentOptions = PaymentOptions(projectID: pluginPaymentOptions.paymentInfo.projectId,
                                            paymentID: pluginPaymentOptions.paymentInfo.paymentId,
                                            paymentAmount: pluginPaymentOptions.paymentInfo.paymentAmount,
                                            paymentCurrency: pluginPaymentOptions.paymentInfo.paymentCurrency,
                                            paymentDescription: pluginPaymentOptions.paymentInfo.paymentDescription,
                                            customerID: pluginPaymentOptions.paymentInfo.customerId,
                                            regionCode: pluginPaymentOptions.paymentInfo.regionCode)
        
        paymentOptions.action = pluginPaymentOptions.actionType.map()
        paymentOptions.receiptData = pluginPaymentOptions.paymentInfo.receiptData
        paymentOptions.languageCode = pluginPaymentOptions.paymentInfo.languageCode
        paymentOptions.forcePaymentMethod = pluginPaymentOptions.paymentInfo.forcePaymentMethod
        paymentOptions.hideSavedWallets = pluginPaymentOptions.paymentInfo.hideSavedWallets ?? false
        paymentOptions.signature = pluginPaymentOptions.paymentInfo.signature
        paymentOptions.additionalFields = pluginPaymentOptions.additionalFields?.map({ field in
            AdditionalField(customName: field.type, value: field.value)
        })
        paymentOptions.mockModeType = pluginPaymentOptions.mockModeType.map()
        paymentOptions.screenDisplayModes = Set(pluginPaymentOptions.screenDisplayModes?.map({ mode in
            mode.map()
        }) ?? [] )
        paymentOptions.hideSavedWallets = pluginPaymentOptions.hideScanningCards ?? false
        paymentOptions.hideSavedWallets = pluginPaymentOptions.paymentInfo.hideSavedWallets ?? false
        paymentOptions.isDarkThemeOn = pluginPaymentOptions.isDarkTheme ?? false
        paymentOptions.brandColor = pluginPaymentOptions.brandColor != nil ? UIColor(hex: pluginPaymentOptions.brandColor!) : nil
        paymentOptions.storedCardType = pluginPaymentOptions.storedCardType != nil ? NSNumber(value: pluginPaymentOptions.storedCardType!) : nil
        
        
        if let recurrentData  = pluginPaymentOptions.recurrentData {
            paymentOptions.recurrentInfo = RecurrentInfo(
                register: recurrentData.register,
                type: recurrentData.type != nil ? RecurrentType(rawValue: recurrentData.type!) : nil,
                expiryDay: recurrentData.expiryDay,
                expiryMonth: recurrentData.expiryMonth,
                expiryYear: recurrentData.expiryYear,
                period: recurrentData.period != nil ? RecurrentPeriod(rawValue: recurrentData.period!) : nil,
                interval: recurrentData.interval,
                time: recurrentData.time,
                startDate: recurrentData.startDate,
                scheduledPaymentID: recurrentData.scheduledPaymentID,
                schedule: recurrentData.schedule?.map({ item in
                    RecurrentInfoSchedule(date: item.date, amount: item.amount)
                })
             )
        }
        
        
        if let applePayCountryCode = pluginPaymentOptions.applePayCountryCode,
           let applePayMerchantId = pluginPaymentOptions.applePayMerchantId,
           let applePayDescription = pluginPaymentOptions.applePayDescription {
            paymentOptions.applePayOptions =  PaymentOptions.ApplePayOptions(applePayMerchantID: applePayMerchantId,
                                                                             applePayDescription: applePayDescription,
                                                                             countryCode: applePayCountryCode)
        }
        
        
        if let recipientInfo  = pluginPaymentOptions.recipientInfo {
            paymentOptions.recipientInfo = RecipientInfo(
                walletId: recipientInfo.walletId,
                walletOwner: recipientInfo.walletOwner,
                pan: recipientInfo.pan,
                cardHolder: recipientInfo.cardHolder,
                country: recipientInfo.country,
                stateCode: recipientInfo.stateCode,
                city: recipientInfo.city,
                address: recipientInfo.address
            )
        }
        
        ecompaySDK.presentPayment(at: controller, paymentOptions: paymentOptions) { sdkResult in
            
            let pluginResult = PluginResult(
                resultCode: sdkResult.status.rawValue,
                payment: sdkResult.payment != nil ? PluginPayment(
                    id: sdkResult.payment!.id ?? "",
                    status: sdkResult.payment!.status ?? "",
                    paymentMethodType: sdkResult.payment!.method ?? "",
                    sum: sdkResult.payment!.sum,
                    type: sdkResult.payment!.type,
                    token: sdkResult.payment!.token,
                    currency: sdkResult.payment!.currency,
                    paymentMassage: nil
                ) : nil,
                errorCode: sdkResult.error?.codeString,
                errorMessage: sdkResult.error?.message
            )
            
            let jsonData = try! JSONEncoder().encode(pluginResult)
            let json = String(data: jsonData, encoding: String.Encoding.utf8)

            result(json)
        }
        
    }
    

}

File: android/src/main/kotlin/com/ecommpay/flutterplugin/EcmpPlugin.kt
package com.ecommpay.flutterplugin

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.annotation.NonNull
import com.ecommpay.flutterplugin.models.PluginPayment
import com.ecommpay.flutterplugin.models.PluginPaymentInfo
import com.ecommpay.flutterplugin.models.PluginPaymentOptions
import com.ecommpay.flutterplugin.models.PluginResult
import com.ecommpay.msdk.ui.EcmpActionType
import com.ecommpay.msdk.ui.EcmpAdditionalFieldType
import com.ecommpay.msdk.ui.EcmpPaymentSDK
import com.ecommpay.msdk.ui.EcmpRecipientInfo
import com.ecommpay.msdk.ui.EcmpRecurrentData
import com.ecommpay.msdk.ui.EcmpRecurrentDataSchedule
import com.ecommpay.msdk.ui.EcmpScreenDisplayMode
import com.ecommpay.msdk.ui.paymentOptions
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import kotlinx.serialization.json.Json
import java.lang.ref.WeakReference

/** EcmpPlugin */
class EcmpPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    PluginRegistry.ActivityResultListener {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private var activityReference = WeakReference<Activity>(null)
    private var contextReference = WeakReference<Context>(null)
    private var result: Result? = null
    private val serializer = Json {
        isLenient = true
        ignoreUnknownKeys = true
        allowSpecialFloatingPointValues = true
        explicitNulls = false
        encodeDefaults = true
    }

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "ecmpplugin")
        channel.setMethodCallHandler(this)
        contextReference = WeakReference(flutterPluginBinding.applicationContext)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityReference = WeakReference(binding.activity)
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityReference.clear()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityReference = WeakReference(binding.activity)
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activityReference.clear()
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        val json = call.arguments as String
        if (call.method == "sdkRun" && json.isNotEmpty()) {
            sdkRun(json = json, result = result)
        } else if (call.method == "getParamsForSignature" && json.isNotEmpty()) {
            getParamsForSignature(json = json, result = result)
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun getParamsForSignature(json: String, result: Result) {
        val pluginPaymentInfo =
            serializer.decodeFromString(PluginPaymentInfo.serializer(), json)
        val ecmpPaymentInfo = pluginPaymentInfo.map()
        val paramsForSignature = ecmpPaymentInfo.getParamsForSignature()
        result.success(paramsForSignature)
    }

    private fun sdkRun(json: String, result: Result) {
        val pluginPaymentOptions =
            serializer.decodeFromString(PluginPaymentOptions.serializer(), json)
        //1. Create EcmpPaymentInfo object
        val ecmpPaymentInfo = pluginPaymentOptions.paymentInfo.map()
        //2. Configure SDK
        val paymentOptions = paymentOptions {
            paymentInfo = ecmpPaymentInfo
            actionType = EcmpActionType.valueOf(pluginPaymentOptions.actionType.name)
            //GooglePay options
            isTestEnvironment = pluginPaymentOptions.googleIsTestEnvironment ?: true
            merchantId = pluginPaymentOptions.googleMerchantId ?: ""
            merchantName = pluginPaymentOptions.googleMerchantName ?: ""
            additionalFields {
                pluginPaymentOptions.additionalFields?.forEach { field ->
                    field {
                        type = EcmpAdditionalFieldType.entries.find { it.value == field.type }
                        value = field.value
                    }
                }
            }
            screenDisplayModes {
                pluginPaymentOptions.screenDisplayModes?.forEach {
                    mode(EcmpScreenDisplayMode.valueOf(it.name))
                }
            }
            if (pluginPaymentOptions.recipientInfo != null)
                recipientInfo = EcmpRecipientInfo(
                    walletOwner = pluginPaymentOptions.recipientInfo.walletOwner,
                    walletId = pluginPaymentOptions.recipientInfo.walletId,
                    country = pluginPaymentOptions.recipientInfo.country,
                    pan = pluginPaymentOptions.recipientInfo.pan,
                    cardHolder = pluginPaymentOptions.recipientInfo.cardHolder,
                    address = pluginPaymentOptions.recipientInfo.address,
                    city = pluginPaymentOptions.recipientInfo.city,
                    stateCode = pluginPaymentOptions.recipientInfo.stateCode,
                )
            if (pluginPaymentOptions.recurrentData != null)
                recurrentData = EcmpRecurrentData(
                    register = pluginPaymentOptions.recurrentData.register,
                    type = pluginPaymentOptions.recurrentData.type?.value,
                    expiryDay = pluginPaymentOptions.recurrentData.expiryDay,
                    expiryMonth = pluginPaymentOptions.recurrentData.expiryMonth,
                    expiryYear = pluginPaymentOptions.recurrentData.expiryYear,
                    period = pluginPaymentOptions.recurrentData.period?.value,
                    interval = pluginPaymentOptions.recurrentData.interval,
                    time = pluginPaymentOptions.recurrentData.time,
                    startDate = pluginPaymentOptions.recurrentData.startDate,
                    scheduledPaymentID = pluginPaymentOptions.recurrentData.scheduledPaymentID,
                    amount = pluginPaymentOptions.recurrentData.amount,
                    schedule = pluginPaymentOptions.recurrentData.schedule?.map {
                        EcmpRecurrentDataSchedule(
                            date = it.date,
                            amount = it.amount
                        )
                    }

                )
            //Parameter to enable hiding or displaying scanning cards feature
            hideScanningCards = pluginPaymentOptions.hideScanningCards ?: false
            isDarkTheme = pluginPaymentOptions.isDarkTheme
        }

        contextReference.get()?.let { context ->
            //4. Create sdk object
            val sdk = EcmpPaymentSDK(
                context = context,
                paymentOptions = paymentOptions,
                mockModeType = EcmpPaymentSDK.EcmpMockModeType.valueOf(pluginPaymentOptions.mockModeType.name)
            )
            this.result = result
            activityReference.get()?.startActivityForResult(sdk.intent, 100)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        when (resultCode) {
            EcmpPaymentSDK.RESULT_SUCCESS -> {
                val paymentJson = data?.getStringExtra(EcmpPaymentSDK.EXTRA_PAYMENT)
                result?.success(
                    serializer.encodeToString(
                        PluginResult.serializer(),
                        PluginResult(
                            resultCode = resultCode,
                            payment = paymentJson?.let {
                                serializer.decodeFromString<PluginPayment>(it)
                            }
                        )
                    )
                )
            }

            EcmpPaymentSDK.RESULT_ERROR -> {
                val errorCode = data?.getStringExtra(EcmpPaymentSDK.EXTRA_ERROR_CODE)
                val errorMessage = data?.getStringExtra(EcmpPaymentSDK.EXTRA_ERROR_MESSAGE)
                result?.success(
                    serializer.encodeToString(
                        PluginResult.serializer(),
                        PluginResult(
                            resultCode = resultCode,
                            errorCode = errorCode,
                            errorMessage = errorMessage
                        )
                    )
                )
            }

            else -> {
                result?.success(
                    serializer.encodeToString(
                        PluginResult.serializer(),
                        PluginResult(resultCode = resultCode)
                    )
                )
            }
        }

        result = null
        return false
    }
}
